File: php/src/EPostak.php
<?php

declare(strict_types=1);

namespace EPostak;

use EPostak\RateLimit;
use EPostak\Resources\Auth;
use EPostak\Resources\Audit;
use EPostak\Resources\Documents;
use EPostak\Resources\Firms;
use EPostak\Resources\Inbound;
use EPostak\Resources\Integrator;
use EPostak\Resources\OAuth;
use EPostak\Resources\Outbound;
use EPostak\Resources\Peppol;
use EPostak\Resources\Webhooks;
use EPostak\Resources\Reporting;
use EPostak\Resources\Extract;
use EPostak\Resources\Account;
use GuzzleHttp\Client as GuzzleClient;
use GuzzleHttp\Exception\GuzzleException;

/**
 * ePošťák API client.
 *
 * @property-read Auth $auth OAuth token mint/renew/revoke + key introspection, rotation, IP allowlist
 * @property-read Audit $audit Per-firm audit feed (cursor-paginated)
 * @property-read Documents $documents Send and receive documents via Peppol
 * @property-read Firms $firms Manage client firms (integrator keys)
 * @property-read Inbound $inbound Pull API — receive and acknowledge inbound Peppol documents
 * @property-read Outbound $outbound Pull API — list sent documents and stream delivery events
 * @property-read Peppol $peppol SMP lookup and Peppol directory search
 * @property-read Webhooks $webhooks Manage webhook subscriptions and pull queue
 * @property-read Reporting $reporting Document statistics and reports
 * @property-read Extract $extract AI-powered OCR extraction from PDFs and images
 * @property-read Account $account Account and firm information
 * @property-read Integrator $integrator Integrator-aggregate endpoints (sk_int_* keys)
 */
class EPostak
{
    private const DEFAULT_BASE_URL = 'https://epostak.sk/api/v1';
    private const DEFAULT_PUBLIC_BASE_URL = 'https://epostak.sk/api';

    private string $clientId;
    private string $clientSecret;
    private string $baseUrl;
    private ?string $firmId;
    private int $maxRetries;

    public Auth $auth;
    public Audit $audit;
    public Documents $documents;
    public Firms $firms;
    public Inbound $inbound;
    public Outbound $outbound;
    public Peppol $peppol;
    public Webhooks $webhooks;
    public Reporting $reporting;
    public Extract $extract;
    public Account $account;
    public Integrator $integrator;

    private HttpClient $http;

    /**
     * Create a new ePošťák API client.
     *
     * @param array{clientId: string, clientSecret: string, baseUrl?: string, firmId?: string, maxRetries?: int} $config Configuration array.
     *   - `clientId`     (required) OAuth client ID (`sk_live_*` or `sk_int_*`).
     *   - `clientSecret` (required) OAuth client secret.
     *   - `baseUrl`      (optional) Override the API base URL (default: https://epostak.sk/api/v1).
     *   - `firmId`       (optional) Scope all requests to this firm ID (required for `sk_int_*`).
     *   - `maxRetries`   (optional) Maximum retries on 429/5xx responses (default: 3).
     *
     * @throws \InvalidArgumentException If clientId or clientSecret is missing.
     *
     * @example
     *   $client = new EPostak(['clientId' => 'sk_live_...', 'clientSecret' => '...']);
     *   $client->documents->send([...]);
     *
     * @example Scoped to a specific firm:
     *   $client = new EPostak(['clientId' => 'sk_int_...', 'clientSecret' => '...', 'firmId' => 'firm_abc']);
     */
    public function __construct(array $config)
    {
        if (empty($config['clientId']) || !is_string($config['clientId'])) {
            throw new \InvalidArgumentException('EPostak: clientId is required');
        }
        if (empty($config['clientSecret']) || !is_string($config['clientSecret'])) {
            throw new \InvalidArgumentException('EPostak: clientSecret is required');
        }

        $this->clientId = $config['clientId'];
        $this->clientSecret = $config['clientSecret'];
        $this->baseUrl = $config['baseUrl'] ?? self::DEFAULT_BASE_URL;
        $this->firmId = $config['firmId'] ?? null;
        $this->maxRetries = $config['maxRetries'] ?? 3;

        $tokenManager = new TokenManager(
            $this->clientId,
            $this->clientSecret,
            $this->baseUrl,
            $this->firmId
        );

        $this->http = new HttpClient($this->baseUrl, $tokenManager, $this->firmId, $this->maxRetries);

        $this->auth = new Auth($this->http);
        $this->audit = new Audit($this->http);
        $this->documents = new Documents($this->http);
        $this->firms = new Firms($this->http);
        $this->inbound = new Inbound($this->http);
        $this->outbound = new Outbound($this->http);
        $this->peppol = new Peppol($this->http);
        $this->webhooks = new Webhooks($this->http);
        $this->reporting = new Reporting($this->http);
        $this->extract = new Extract($this->http);
        $this->account = new Account($this->http);
        $this->integrator = new Integrator($this->http);
    }

    /**
     * Return the rate-limit snapshot from the most recent API response.
     *
     * Populated after every successful (or failed) call that reaches the server.
     * Returns `null` before the first request or when the last response did not
     * include `X-RateLimit-*` headers.
     *
     * @return RateLimit|null
     *
     * @example
     *   $client->documents->list();
     *   $rl = $client->getLastRateLimit();
     *   if ($rl !== null) {
     *       echo "Remaining: {$rl->remaining}/{$rl->limit}";
     *   }
     */
    public function getLastRateLimit(): ?RateLimit
    {
        return $this->http->getLastRateLimit();
    }

    /**
     * Create a new client instance scoped to a specific firm.
     *
     * Useful when an integrator key needs to switch between client firms.
     * Returns a fresh EPostak instance; the original is not modified.
     *
     * @param string $firmId The firm ID to scope requests to.
     * @return self A new client instance scoped to the given firm.
     *
     * @example
     *   $firmClient = $client->withFirm('firm_abc');
     *   $firmClient->documents->send([...]);
     */
    public function withFirm(string $firmId): self
    {
        return new self([
            'clientId' => $this->clientId,
            'clientSecret' => $this->clientSecret,
            'baseUrl' => $this->baseUrl,
            'firmId' => $firmId,
            'maxRetries' => $this->maxRetries,
        ]);
    }

    /**
     * Validate a UBL XML document against the Peppol BIS 3.0 3-layer rules.
     *
     * This endpoint is PUBLIC -- no API key is required. Rate-limited to
     * 20 requests per minute per IP. Can be called statically without
     * constructing an EPostak client, which makes it suitable for CLI
     * tools or test harnesses.
     *
     * @param string      $xml     UBL 2.1 XML invoice or credit note.
     * @param string|null $baseUrl Optional override for the public API base URL
     *                             (defaults to https://epostak.sk/api).
     * @return array Full 3-layer Peppol BIS 3.0 validation report.
     * @throws EPostakError On non-2xx responses or network errors.
     *
     * @example
     *   $report = \EPostak\EPostak::validate(file_get_contents('invoice.xml'));
     *   var_dump($report['valid']);
     */
    public static function validate(string $xml, ?string $baseUrl = null): array
    {
        $base = rtrim($baseUrl ?? self::DEFAULT_PUBLIC_BASE_URL, '/');
        $client = new GuzzleClient([
            'base_uri' => $base . '/',
            'http_errors' => false,
        ]);

        try {
            $response = $client->request('POST', 'validate', [
                'headers' => [
                    'Content-Type' => 'application/xml',
                    'Accept' => 'application/json',
                ],
                'body' => $xml,
            ]);
        } catch (GuzzleException $e) {
            throw new EPostakError(0, ['error' => $e->getMessage()]);
        }

        $statusCode = $response->getStatusCode();
        $body = $response->getBody()->getContents();

        if ($statusCode >= 400) {
            $decoded = [];
            try {
                $decoded = json_decode($body, true) ?? [];
            } catch (\Throwable) {
                $decoded = ['error' => $response->getReasonPhrase()];
            }
            throw new EPostakError($statusCode, $decoded, $response->getHeaders());
        }

        return $body === '' ? [] : (json_decode($body, true) ?? []);
    }

}

File: ruby/lib/epostak.rb
# frozen_string_literal: true

require_relative "epostak/version"
require_relative "epostak/error"
require_relative "epostak/token_manager"
require_relative "epostak/http_client"
require_relative "epostak/webhook_signature"
require_relative "epostak/oauth"
require_relative "epostak/client"
require_relative "epostak/resources/auth"
require_relative "epostak/resources/audit"
require_relative "epostak/resources/documents"
require_relative "epostak/resources/inbox"
require_relative "epostak/resources/firms"
require_relative "epostak/resources/integrator"
require_relative "epostak/resources/peppol"
require_relative "epostak/resources/peppol_directory"
require_relative "epostak/resources/webhooks"
require_relative "epostak/resources/webhook_queue"
require_relative "epostak/resources/reporting"
require_relative "epostak/resources/account"
require_relative "epostak/resources/extract"

# Top-level namespace for the ePošťák Ruby SDK.
#
# @example Quick start
#   client = EPostak::Client.new(client_id: "sk_live_xxxxx", client_secret: "secret")
#   result = client.documents.send_document(
#     receiverPeppolId: "0245:1234567890",
#     items: [{ description: "Consulting", quantity: 10, unitPrice: 100, vatRate: 23 }]
#   )
#
# @example Verify a webhook
#   result = EPostak.verify_webhook_signature(
#     payload: request.body.read,
#     signature_header: request.env["HTTP_X_EPOSTAK_SIGNATURE"].to_s,
#     secret: ENV["EPOSTAK_WEBHOOK_SECRET"]
#   )
#   halt 400 unless result[:valid]
module EPostak
  # Default base URL for the ePošťák API.
  DEFAULT_BASE_URL = "https://epostak.sk/api/v1"

  # Default base URL for the public (non-authenticated) ePošťák API.
  # Used by the public {.validate} endpoint which requires no API key.
  DEFAULT_PUBLIC_BASE_URL = "https://epostak.sk/api"

  # Validate a UBL XML document against the Peppol BIS 3.0 3-layer rules.
  #
  # This endpoint is *public* — no API key is required. Rate-limited to
  # 20 requests per minute per IP.
  #
  # @param xml [String] UBL 2.1 XML invoice or credit note as a string
  # @param base_url [String, nil] Optional override for the public API base
  #   URL. Defaults to +https://epostak.sk/api+.
  # @return [Hash] Full 3-layer Peppol BIS 3.0 validation report
  # @raise [EPostak::Error] On non-2xx responses
  #
  # @example
  #   report = EPostak.validate(File.read("invoice.xml"))
  #   puts "Valid!" if report["valid"]
  def self.validate(xml, base_url: nil)
    require "faraday"
    require "json"

    url = base_url || DEFAULT_PUBLIC_BASE_URL
    conn = Faraday.new(url: url)
    response = conn.post("/validate") do |req|
      req.headers["Content-Type"] = "application/xml"
      req.body = xml
    end

    unless response.success?
      body =
        begin
          JSON.parse(response.body)
        rescue StandardError
          { "error" => response.reason_phrase || "validate request failed" }
        end
      raise EPostak.build_api_error(response.status, body, response.headers)
    end

    return nil if response.body.nil? || response.body.empty?

    begin
      JSON.parse(response.body)
    rescue JSON::ParserError
      response.body
    end
  end
end
